pkg/policy/mutate/validate.go

```diff
@@ -1,63 +1,54 @@
 package mutate
 
 import (
-	"errors"
 	"fmt"
 
 	kyverno "github.com/kyverno/kyverno/api/kyverno/v1"
-	commonAnchors "github.com/kyverno/kyverno/pkg/engine/anchor/common"
-	"github.com/kyverno/kyverno/pkg/policy/common"
 )
 
 // Mutate provides implementation to validate 'mutate' rule
 type Mutate struct {
-	// rule to hold 'mutate' rule specifications
-	rule kyverno.Mutation
+	mutation kyverno.Mutation
 }
 
 //NewMutateFactory returns a new instance of Mutate validation checker
-func NewMutateFactory(rule kyverno.Mutation) *Mutate {
-	m := Mutate{
-		rule: rule,
+func NewMutateFactory(m kyverno.Mutation) *Mutate {
+	return &Mutate{
+		mutation: m,
 	}
-	return &m
 }
 
 //Validate validates the 'mutate' rule
 func (m *Mutate) Validate() (string, error) {
-	rule := m.rule
-	// JSON Patches
-	if len(rule.Patches) != 0 {
-		for i, patch := range rule.Patches {
-			if err := validatePatch(patch); err != nil {
-				return fmt.Sprintf("patch[%d]", i), err
-			}
-		}
-	}
-	// Overlay
-	if rule.Overlay != nil {
-		path, err := common.ValidatePattern(rule.Overlay, "/", []commonAnchors.IsAnchor{commonAnchors.IsConditionAnchor, commonAnchors.IsAddingAnchor})
-		if err != nil {
-			return path, err
-		}
+	if m.hasForEach() {
+		return m.validateForEach()
 	}
+
 	return "", nil
 }
 
-// Validate if all mandatory PolicyPatch fields are set
-func validatePatch(pp kyverno.Patch) error {
-	if pp.Path == "" {
-		return errors.New("JSONPatch field 'path' is mandatory")
+func (m *Mutate) validateForEach() (string, error) {
+	if m.hasPatchStrategicMerge() || m.hasPatchesJSON6902() {
+		return "foreach", fmt.Errorf("mutate rule must contain either a `foreach`, a `patchStrategicMerge`, or a `patchesJson6902` declaration")
```

Comment:
For this case the rule has either `patchStrategicMerge` or `patchesJson6902`, why it's an error?

Reply:
This check is for a `forEach`: If the rule has a `forEach` it cannot have a `patchStrategicMerge` or `patchesJson6902`.

Reply:
Understood. The message seems confusing after I read the code, can we change to "patchStrategicMerge and patchesJson6902 cannot be defined with foreach" or similar?

Reply:
Ok. How about, "only one of `foreach`, `patchStrategicMerge`, or `patchesJson6902` is allowed"?

Reply:
Sounds good!